### PR TITLE
Handle index/source not found and timeout/transport errors

### PR DESCRIPTION
**Pull request by guilload**
### Description
- Split `GetOrCreateOpenShardsReponse` into `successes` and `failures` to keep track of indexes/sources not found
- Remove leaders from ingester pool on transport error

### How was this PR tested?
- Added unit tests
- Tested locally

### CODE REVIEW

File: quickwit/quickwit-proto/protos/quickwit/control_plane.proto [hunk 1]
```diff
@@ -80,13 +81,27 @@ message ClosedShards {
...
-    repeated quickwit.ingest.Shard open_shards = 3;
+message GetOrCreateOpenShardsSuccess {
+  uint32 subrequest_id = 1;
+  string index_uid = 2;
+  string source_id = 3;
+  repeated quickwit.ingest.Shard open_shards = 4;
+}
+
+enum GetOrCreateOpenShardsFailureReason {
+  GET_OR_CREATE_OPEN_SHARDS_FAILURE_REASON_UNSPECIFIED = 0;
```
Comment (fulmicoton) [inline quickwit/quickwit-proto/protos/quickwit/control_plane.proto:97]: Hmm. Repeating the prefix might not be useful here? 
When dealing with enum in rust, we rarely `use GetOrCreateOpneShardsFailureReason::*;`
What do you think?

File: quickwit/quickwit-proto/protos/quickwit/metastore.proto [hunk 2]
```diff
@@ -263,22 +263,24 @@ message OpenShardsRequest {
 }
 
 message OpenShardsSubrequest {
-    string index_uid = 1;
-    string source_id = 2;
-    string leader_id = 3;
-    optional string follower_id = 4;
-    uint64 next_shard_id = 5;
+  uint32 subrequest_id = 1;
```
Comment (fulmicoton) [inline quickwit/quickwit-proto/protos/quickwit/metastore.proto:266]: weird indentation?

File: quickwit/quickwit-proto/protos/quickwit/control_plane.proto (hunk 1 above)
Comment (fulmicoton) [reply quickwit/quickwit-proto/protos/quickwit/control_plane.proto:97]: on further thoughts that's specific to rust and your naming is the right one.

File: quickwit/quickwit-ingest/src/ingest_v2/workbench.rs [hunk 3]
```diff
@@ -29,14 +32,17 @@ use tracing::warn;
...
 /// A helper struct for managing the state of the subrequests of an ingest request during multiple
 /// persist attempts.
+#[derive(Default)]
 pub(super) struct IngestWorkbench {
-    subworkbenches: HashMap<SubrequestId, IngestSubworkbench>,
-    num_successes: usize,
+    pub subworkbenches: HashMap<SubrequestId, IngestSubworkbench>,
+    pub num_successes: usize,
+    pub num_attempts: usize,
+    pub max_num_attempts: usize,
```
Comment (fulmicoton) [inline quickwit/quickwit-ingest/src/ingest_v2/workbench.rs:40]: Can we add a comment explaining this si an overall number of attempts. Not a per sub request number.

File: quickwit/quickwit-ingest/src/ingest_v2/workbench.rs [hunk 4]
```diff
@@ -48,22 +54,77 @@ impl IngestWorkbench {
...
+        else {
+            warn!(
+                "could not find subrequest `{}` in workbench",
+                open_shards_failure.subrequest_id
+            );
+            return;
+        };
+        subworkbench.num_attempts += 1;
+
+        let last_failure_opt = match open_shards_failure.reason() {
```
Comment (fulmicoton) [inline quickwit/quickwit-ingest/src/ingest_v2/workbench.rs:109]:
```suggestion
        // Some failure are not retryable. In that case we record them
        // as the last failure reason and the associated subrequest will be ignore by the retry process.
        let last_failure_opt = match open_shards_failure.reason() {
```

File: quickwit/quickwit-ingest/src/ingest_v2/workbench.rs [hunk 5]
```diff
@@ -48,22 +54,77 @@ impl IngestWorkbench {
...
+            }
+            GetOrCreateOpenShardsFailureReason::SourceNotFound => {
+                Some(SubworkbenchFailure::SourceNotFound)
+            }
+            GetOrCreateOpenShardsFailureReason::Unspecified => {
+                warn!(
+                    "failure reason for subrequest `{}` is unspecified",
+                    open_shards_failure.subrequest_id
+                );
+                None
```
Comment (fulmicoton) [inline quickwit/quickwit-ingest/src/ingest_v2/workbench.rs:121]: I don't understand why we record `None` here. 
We seem to have logic to different retryable (you called that transient) from non retryable errors.

Could we reserve `None` to we have never received a failure yet?

**Review by fulmicoton — APPROVED** [review]

File: quickwit/quickwit-proto/protos/quickwit/control_plane.proto (hunk 1 above)
Comment (guilload) [reply quickwit/quickwit-proto/protos/quickwit/control_plane.proto:97]: Discussed offline.

The [best practice for Protobuf](https://protobuf.dev/programming-guides/dos-donts/#unspecified-enum) enums is to reuse the name of the enum as a prefix for the variant and always have `FOO_UNSPECIFIED` for the `0` tag.

Prost does the right when code-generating the corresponding Rust enums and strips the common prefix.

**Pull request merged**